### PR TITLE
HDFS-13693: Remove unnecessary search in INodeDirectory.addChild during image loa…

### PR DESCRIPTION
**Pull request by leosunli**
…ding

Signed-off-by: sunlisheng <sunlisheng@xiaomi.com>

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 30 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1141 | trunk passed |
| +1 | compile | 62 | trunk passed |
| +1 | checkstyle | 46 | trunk passed |
| +1 | mvnsite | 68 | trunk passed |
| +1 | shadedclient | 781 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 53 | trunk passed |
| 0 | spotbugs | 163 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 161 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 59 | the patch passed |
| +1 | compile | 55 | the patch passed |
| +1 | javac | 55 | the patch passed |
| -0 | checkstyle | 39 | hadoop-hdfs-project/hadoop-hdfs: The patch generated 1 new + 28 unchanged - 0 fixed = 29 total (was 28) |
| +1 | mvnsite | 62 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 743 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 50 | the patch passed |
| +1 | findbugs | 165 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 5078 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 31 | The patch does not generate ASF License warnings. |
| | | 8697 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.TestDFSShell |
|   | hadoop.hdfs.TestErasureCodingPolicies |
|   | hadoop.hdfs.web.TestWebHdfsTimeouts |
|   | hadoop.hdfs.server.blockmanagement.TestBlockStatsMXBean |
|   | hadoop.hdfs.TestDFSStripedOutputStream |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1040 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux c0a819cb6f74 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 1e727cf |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/1/artifact/out/diff-checkstyle-hadoop-hdfs-project_hadoop-hdfs.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/1/testReport/ |
| Max. process+thread count | 4280 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 32 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1085 | trunk passed |
| +1 | compile | 62 | trunk passed |
| +1 | checkstyle | 47 | trunk passed |
| +1 | mvnsite | 71 | trunk passed |
| +1 | shadedclient | 781 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 53 | trunk passed |
| 0 | spotbugs | 166 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 164 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 62 | the patch passed |
| +1 | compile | 58 | the patch passed |
| +1 | javac | 58 | the patch passed |
| +1 | checkstyle | 40 | the patch passed |
| +1 | mvnsite | 64 | the patch passed |
| +1 | whitespace | 1 | The patch has no whitespace issues. |
| +1 | shadedclient | 738 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 48 | the patch passed |
| +1 | findbugs | 172 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 4965 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 33 | The patch does not generate ASF License warnings. |
| | | 8540 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.datanode.TestDirectoryScanner |
|   | hadoop.hdfs.web.TestWebHdfsTimeouts |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1040 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 0695508fe7a3 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 1e727cf |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/2/testReport/ |
| Max. process+thread count | 5203 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by leosunli** [comment]
Hi @goiri  Could you have time to review this patch? Thank you.

**Comment by goiri** [comment]
I think the patch makes sense.
However, it looks like Wei-Chiu Chuang is taking care of the review in the JIRA.
I would prefer for him to do the final review.

**Comment by leosunli** [comment]
Thank @goiri for your reply. I would be looking for Wei-Chiu Chuang to help review it.

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/INodeDirectory.java [hunk 1]
```diff
@@ -572,6 +572,25 @@ public boolean addChild(INode node) {
...
+   * During image loading, the search is unnecessary since the insert position
+   * should always be at the end of the map given the sequence they are
+   * serialized on disk.
+   */
+  public boolean addChildAtLoading(INode node) {
+    int pos;
+    if (!node.isReference()) {
+      pos = (children == null) ? (-1) : (-children.size() - 1);
+    } else {
+      pos = searchChildren(node.getLocalNameBytes());
```
Comment (jojochuang) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/INodeDirectory.java]: can we reuse addChild() here to make it cleaner?
My understanding is that in the case of INodeReference, we fall back to the original behavior.

That is, 
`
} else {
  return addChild(node);
}
`

**Comment by leosunli** [comment]
Thank @jojochuang  for your review. I have updated this patch. Could you have time to review it?

**Comment by jojochuang** [comment]
+1

**Comment by leosunli** [comment]
hi @jojochuang If you think this patch no problem, could you help merge it to trunk ? Thank you.

**Comment by leosunli** [comment]
hi @goiri  @jojochuang have +1 for this patch.  He may not have time to merge to trunk. Could you have time to help do it? Thank you.

**Comment by goiri** [comment]
The parallel life of JIRAs and PRs is driving me a little crazy.
We have both the patch and the diff here and then we also have comments in both.

Anyway, He Xiaoqiao seems to have comments in the JIRA.
It would be good to get his +1.

**Comment by Hexiaoqiao** [comment]
Sorry for notice this PR untimely. Thanks @leosunli for your contribution.
+1 from me for HDFS-13693-005.patch at [HDFS-13693|https://issues.apache.org/jira/browse/HDFS-13693]. It is little different between these two patches. I would strongly advise to unify them.

> The parallel life of JIRAs and PRs is driving me a little crazy.
+1. BTW, It seems a few PRs both submit to here and JIRA. I am confused which one is community recommendation.

**Comment by leosunli** [comment]
Thank @Hexiaoqiao for your comment. I have unified this PR and this JIAR and updated this PR. Thank you again.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 38 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1052 | trunk passed |
| +1 | compile | 58 | trunk passed |
| +1 | checkstyle | 39 | trunk passed |
| +1 | mvnsite | 68 | trunk passed |
| +1 | shadedclient | 724 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 51 | trunk passed |
| 0 | spotbugs | 167 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 164 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 59 | the patch passed |
| +1 | compile | 54 | the patch passed |
| +1 | javac | 54 | the patch passed |
| +1 | checkstyle | 37 | the patch passed |
| +1 | mvnsite | 62 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 705 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 44 | the patch passed |
| +1 | findbugs | 165 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 5029 | hadoop-hdfs in the patch failed. |
| +1 | asflicense | 35 | The patch does not generate ASF License warnings. |
| | | 8440 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.web.TestWebHdfsTimeouts |
|   | hadoop.hdfs.server.namenode.TestNameNodeMetadataConsistency |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |
|   | hadoop.hdfs.server.namenode.ha.TestBootstrapAliasmap |
|   | hadoop.hdfs.qjournal.server.TestJournalNodeSync |
|   | hadoop.hdfs.server.datanode.fsdataset.impl.TestFsDatasetImpl |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.8 Server=18.09.8 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1040 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux edd457aefc14 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 4e66cb9 |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/3/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/3/testReport/ |
| Max. process+thread count | 4830 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project/hadoop-hdfs |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by leosunli** [comment]
> It seems a few PRs both submit to here and JIRA. I am confused which one is community recommendation.
In fact, I have been very troubled by this problem.So I have always been on both sides.
I think PR is more convenient to review. Thank you.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 0 | Docker mode activated. |
| -1 | patch | 12 | https://github.com/apache/hadoop/pull/1040 does not apply to trunk. Rebase required? Wrong Branch? See https://wiki.apache.org/hadoop/HowToContribute for help. |


| Subsystem | Report/Notes |
|----------:|:-------------|
| GITHUB PR | https://github.com/apache/hadoop/pull/1040 |
| JIRA Issue | HDFS-13693 |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1040/4/console |
| versions | git=2.7.4 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by jojochuang** [comment]
Close this PR. The commit was merged into trunk. See the corresponding JIRA https://issues.apache.org/jira/browse/HDFS-13693

**Pull request closed without merging**